esmvalcore/iris_helpers.py

```diff
@@ -106,6 +108,74 @@ def date2num(date, unit, dtype=np.float64):
         return dtype(num)
 
 
+def merge_cube_attributes(
+    cubes: Sequence[Cube],
+    delimiter: str = ' ',
+) -> None:
+    """Merge attributes of all given cubes in-place.
+
+    After this operation, the attributes of all given cubes are equal. This is
+    useful for operations that combine cubes, such as
+    :meth:`iris.cube.CubeList.merge_cube` or
+    :meth:`iris.cube.CubeList.concatenate_cube`.
+
+    Note
+    ----
+    This function differs from :func:`iris.util.equalise_attributes` in this
+    respect that it does not delete attributes that are not identical but
+    rather concatenates them (sorted) using the given ``delimiter``. E.g., the
+    attributes ``exp: historical`` and ``exp: ssp585`` end up as ``exp:
+    historical ssp585`` using the default ``delimiter = ' '``.
+
+    Parameters
+    ----------
+    cubes:
+        Input cubes whose attributes will be modified in-place.
+    delimiter:
+        Delimiter that is used to concatenate non-identical attributes.
+
+    """
+    if len(cubes) <= 1:
+        return
+
+    # Step 1: collect all attribute values in a list
+    attributes: Dict[Any, List[Any]] = {}
+    for cube in cubes:
+        for (attr, val) in cube.attributes.items():
+            attributes.setdefault(attr, [])
+            attributes[attr].append(val)
+
+    # Step 2: if values are not equal, first convert them to strings (so that
+    # set() can be used); then extract unique elements from this list, sort it,
+    # and use the delimiter to join all elements to a single string
+    final_attributes: Dict[Any, Any] = {}
+    for (attr, vals) in attributes.items():
+        if _contains_identical_values(vals):
```

Comment:
Very good question. The main reason I am using `np.array_equal` here is that it works well with basically any input type.

#### Example why it's superior to `==`:

```py
if np.array_equal(np.arange(2), 1):
    ...
```

works well, while

```py
if np.arange(2) == 1:
    ...
```

raises an error (`ValueError: The truth value of an array ...`).

#### Example why it's superior to checking the length of a `set()`:

To avoid `TypeError: unhashable type`, I use `str()` on all the elements before creating the sets. This is for example problematic when comparing dictionaries, e.g.

```py
>>> {1: 1, 2:2} == {2:2, 1:1}
True
```

while

```py
>>> str({1: 1, 2:2}) == str({2:2, 1:1})
False
```

I know that using `np.array_equal` is not optimal, but it appeared to me as the best solution. On the other hand I am not sure how often we will encounter dictionaries as attributes values, so I would also be fine with implementing your proposed solution.

Reply:
@zklaus Is that allowed by CF, having a mapping as an attribute value? All I could find was strings, numbers, and 'data': https://cfconventions.org/cf-conventions/cf-conventions.html#attribute-appendix. At first glance, 'data' appears to be mostly lists of numbers.

Reply:
Found this bit of information in the [NetCDF4 documentation](https://unidata.github.io/netcdf4-python/#attributes-in-a-netcdf-file): 

> Attributes can be strings, numbers or sequences.

so this should make it straightforward to compare them using just Python code. I tried it out and numerical sequences are loaded as numpy arrays, while sequences of strings are loaded as Python `list`s. If you convert the sequences to `tuple`s, `set` should work just fine in removing the duplicates.